Add a python script for making releases.

Sopel97 opened the pull request:
This PR adds a python script that allows easy creation of binary releases for stockfish. It should work with a fresh installation of python >=3.6.3. It could even be put somewhere in the cloud if people see fit. I despise CI "scripts" so it's done like this instead.

The way it works is it clones the stockfish reposity (depth=1, one branch, by default the master branch), and for build scheduled it copies it to a temporary directory in which `make` is executed, and the resulting binary is moved to the specified output directory with the name. This way multiple builds can be run in parallel (the maximum amount of parallel builds can be set via `--concurrency` option).

The script is tested on windows with MinGW with GCC 10.1, but it should work on other systems as well.

It is possible to do profile-builds and they will be restarted under Intel SDE (PR pending) if `--sde` is specified and the build fails (due to illegal instructions). Currently under windows it generates a window with error that requires manual confirmation and there is no way I could find to supress it since the binary is executed from the make script. If anyone has a better way of detecting such failure it would be great.

```
usage: make_release.py [-h] [--sde SDE] [--concurrency CONCURRENCY]
                       [--repository REPOSITORY] [--clone-dir CLONE_DIR]
                       [--out-dir OUT_DIR] [--build-type BUILD_TYPE]
                       [--prefix PREFIX] [--verbose] [--keep-logs]
                       [archs [archs ...]]

Stockfish release maker

positional arguments:
  archs                 A list of architectures to compile. If not provided
                        then most sensible archs will be built. (default:
                        ['x86-64-vnni512', 'x86-64-vnni256', 'x86-64-avx512',
                        'x86-64-bmi2', 'x86-64-avx2', 'x86-64-sse41-popcnt',
                        'x86-64-modern', 'x86-64-ssse3',
                        'x86-64-sse3-popcntx86-64', 'x86-32-sse41-popcnt',
                        'x86-32-sse2', 'x86-32', 'general-64', 'general-32'])

optional arguments:
  -h, --help            show this help message and exit
  --sde SDE             Path to Intel SDE. If not specified it will not be
                        used. If specified it will be used as a fallback for
                        architectures that fail to compile. (default: None)
  --concurrency CONCURRENCY
                        The maximum number of jobs to run concurrently.
                        (default: 1)
  --repository REPOSITORY
                        The path to the repository to clone. (default:
                        https://github.com/official-stockfish/Stockfish.git)
  --clone-dir CLONE_DIR
                        The directory to clone the repository to. (default:
                        cloned)
  --out-dir OUT_DIR     The directory that shall cotain the resulting
                        binaries. (default: bin)
  --build-type BUILD_TYPE
                        The build type to use in make. (default: profile-
                        build)
  --prefix PREFIX       The string put in the binary name after stockfish- but
                        before -arch. For example "win" or "linux". Will be
                        ommited if empty. (default: )
  --verbose             Keep output of running programs. (default: False)
  --keep-logs           Whether to keep logs in the output directory.
                        Incompabile with --verbose. (default: False)
```

vondele commented:
The following changes made it work on linux, including profile-build using the sde.

```diff
diff --git a/script/make_release.py b/script/make_release.py
index fd1f1de7e..ef4733725 100644
--- a/script/make_release.py
+++ b/script/make_release.py
@@ -7,6 +7,13 @@ import stat
 import concurrent.futures
 import argparse
 import logging
+import platform
+
+IS_WINDOWS = "windows" in platform.system().lower()
+if IS_WINDOWS:
+   exe_suffix=".exe"
+else:
+   exe_suffix=""
 
 DEFAULT_ARCHS = [
     'x86-64-vnni512',
@@ -17,7 +24,7 @@ DEFAULT_ARCHS = [
     'x86-64-sse41-popcnt',
     'x86-64-modern',
     'x86-64-ssse3',
-    'x86-64-sse3-popcnt'
+    'x86-64-sse3-popcnt',
     'x86-64',
     'x86-32-sse41-popcnt',
     'x86-32-sse2',
@@ -36,7 +43,7 @@ parser.add_argument('--repository', default='https://github.com/official-stockfi
 parser.add_argument('--clone-dir', default='repo', type=str,
                     help='The directory to clone the repository to.')
 parser.add_argument('--out-dir', default='bin', type=str,
-                    help='The directory that shall cotain the resulting binaries.')
+                    help='The directory that shall contain the resulting binaries.')
 parser.add_argument('--build-type', default='profile-build', type=str,
                     help='The build type to use in make.')
 parser.add_argument('--prefix', default='', type=str,
@@ -44,14 +51,14 @@ parser.add_argument('--prefix', default='', type=str,
 parser.add_argument('--verbose', default=False, action='store_true',
                     help='Keep output of running programs.')
 parser.add_argument('--keep-logs', default=False, action='store_true',
-                    help='Whether to keep logs in the output directory. Incompabile with --verbose.')
+                    help='Whether to keep logs in the output directory. Incompatible with --verbose.')
 parser.add_argument('archs', nargs='*', default=DEFAULT_ARCHS, type=str,
                     help='A list of architectures to compile. If not provided then most sensible archs will be built.')
 
 args = parser.parse_args()
 
 if args.verbose and args.keep_logs:
-    parser.error('--verbose and --keep-logs are currently not compatibile.')
+    parser.error('--verbose and --keep-logs are currently not compatible.')
 
 SDE_PATH = args.sde
 NUM_THREADS = args.concurrency
@@ -89,9 +96,11 @@ def remove(path):
         raise ValueError("file {} is not a file or dir.".format(path))
 
 def copytree(src, dst, symlinks=False, ignore=None):
+    os.mkdir(dst)
     for item in os.listdir(src):
         s = os.path.join(src, item)
         d = os.path.join(dst, item)
+        print(s,d)
         if os.path.isdir(s):
             shutil.copytree(s, d, symlinks, ignore)
         else:
@@ -169,8 +178,8 @@ def compile_target(repo_dir, dest_dir, build_type='profile-build', arch='x86-64-
             success = True
 
         if success:
-            src_file = src_dir + 'stockfish.exe'
-            dest_file = os.path.join(dest_dir, 'stockfish{}-{}-{}.exe'.format(prefix, build_type, arch))
+            src_file = src_dir + 'stockfish' + exe_suffix
+            dest_file = os.path.join(dest_dir, 'stockfish{}-{}-{}{}'.format(prefix, build_type, arch, exe_suffix))
 
             log('{}: Finished build. Copying {} to {}.'.format(arch, src_file, dest_file))
```

Sopel97 commented:
I've updated the script with the suggested changes. I also added some comments and improved control flow.

ppigazzini commented:
With python 3.4 or later IMO better to use `pathlib`, see https://docs.python.org/3/library/pathlib.html

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,262 @@
+        with TemporaryDirectory('./stockfish-{}-{}-'.format(build_type, arch)) as copy_dir:
+            log('{}: Created temporary directory {}'.format(arch, copy_dir.path))
+
+            src_dir = copy_dir.path + '/src/'
```
w/o using `pathlib` better to use:
`src_dir = os.path.join(copy_dir.path, '/src/')`

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,262 @@
+IS_WINDOWS = 'windows' in platform.system().lower()
+EXE_SUFFIX = '.exe' if IS_WINDOWS else ''
+
+def log(s):
```
I will prefer to use directly `print()`.

For advanced use there is `logging`, see https://docs.python.org/3.6/howto/logging.html

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,259 @@
+    else:
+        raise ValueError("file {} is not a file or dir.".format(path))
+
+def copytree(src, dst, symlinks=False, ignore=None):
```
IMO this function should be replaced by `shutil.copytree(src, dst)`
https://docs.python.org/3/library/shutil.html#shutil.copytree

Sopel97 replied:
doesn't work if the directory already exists (at least on windows)

Sopel97 replied:
fails when the destination directory already exists. Option to ignore this error only in python >=3.8

ppigazzini replied:
If I'm not wrong the destination directory has a random uuid, so the probability to have already that directory should be very low. And if the destination directory already exists I don't know if the best thing to do is to rewrite it.

Sopel97 replied:
the directory exists during a copy because it's created before that

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,259 @@
+IS_WINDOWS = 'windows' in platform.system().lower()
+EXE_SUFFIX = '.exe' if IS_WINDOWS else ''
+
+def rmtree(top):
```
`shutil.rmtree()`
https://docs.python.org/3/library/shutil.html#shutil.rmtree

Sopel97 replied:
on windows it's needed to change the permissions for the deleted files, shutil.rmtree just doesn't work alone.

ppigazzini replied:
https://docs.python.org/3/library/shutil.html#rmtree-example
```python
import os, stat
import shutil

def remove_readonly(func, path, _):
    "Clear the readonly bit and reattempt the removal"
    os.chmod(path, stat.S_IWRITE)
    func(path)

shutil.rmtree(directory, onerror=remove_readonly)
```

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,259 @@
+    and all its content is deleted
+    Otherwise a ValueError is thrown.
+    '''
+    if not os.path.exists(path):
```
IMO this check should be dropped, `else:` catches in proper way the not existing path.

Sopel97 replied:
I wanted to keep it idempotent in case the file/dir already doesn't exist

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,263 @@
+                print('{}: Error running `{}`. Exit code: {}' \
+                    .format(arch, make_command_log, ret.returncode))
+            else:
+                src_file = src_dir + 'stockfish' + EXE_SUFFIX
```
`src_file = os.path.join(src_dir, 'stockfish' + EXE_SUFFIX)`

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,263 @@
+        tasks = dict()
+        for arch in ARCHS:
+            print('Scheduling ARCH={}'.format(arch))
+            future = executor.submit(lambda: run_log_errors(compile_target, REPO_DIR, OUT_DIR, BUILD_TYPE, arch, PREFIX))
```
The script runs fine also without `lambda:`. It's necessary?

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,263 @@
+
+            print('{}: Copying original files from {} to {}'.format(arch, repo_dir, copy_dir.path))
+
+            copytree(repo_dir, copy_dir.path)
```
Copying only the `src` tree should be simpler/safer (e.g. the read-only files seems to be only in `.git` directory).

`copytree(os.path.join(repo_dir, 'src'), os.path.join(copy_dir.path, 'src'))`

Sopel97 replied:
good point

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,219 @@
+            make_command = ['make', build_type, 'ARCH={}'.format(arch)]
+            make_command_log = 'make {} ARCH={}'.format(build_type, arch)
+            if SDE_PATH:
+                make_command += ['SDE_PATH={}'.format(SDE_PATH)]
```
Add double quotes to deal with spaces in path.
```python
            if SDE_PATH:
                make_command += ['SDE_PATH="{}"'.format(SDE_PATH)]
                make_command_log += ' SDE_PATH="{}"'.format(SDE_PATH)
```

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,219 @@
+
+parser = argparse.ArgumentParser(description='Stockfish release maker', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+parser.add_argument('--sde', default=None, type=str,
+                    help='Path to Intel SDE. If not specified it will not be used. If specified it will be used as a fallback for architectures that fail to compile.')
```
This little change is a life changer for a SDE nooby like me :)
`help='Absolute path to Intel SDE application.`

ppigazzini commented:
It works fine on:
- Windows/msys2: w/ and w/o SDE
- Windows/WSL2 (Ubuntu 20.04): w/o SDE
- Linux/Ubuntu20.04: w/ and w/o SDE

ppigazzini commented:
@Sopel97 on Windows/msys2 w/ a profile-build for a wrong ARCH there is the same cli message error of linux, no windows to be closed by hand.

snicolet commented:
I suggest to rename the variable `REPO_DIR` to `CLONE_DIR` to avoid the semantic clash between `REPO` and `REPO_DIR`. Maybe also the default name for `CLONE_DIR` could be `cloned`

snicolet commented:
It would also be nice to have a two lines description of the script near the top of the file, with maybe two examples:
1. `make_release.py --help`
2. a complete working example

ppigazzini commented:
I haven't any error window on Windows 10 msys2, I'm wondering about your Windows setup.

vondele commented:
as we are at feature requests.... 
1) verifying the bench of the generated binaries
2) packing them up in a .zip file together with a .zip of the corresponding source.

Sopel97 commented:
> 
> 
> I haven't any error window on Windows 10 msys2, I'm wondering about your Windows setup.

win 7

MichaelB7 commented:
> > I haven't any error window on Windows 10 msys2, I'm wondering about your Windows setup.
> 
> win 7, but not running in msys2, might be different under msys
> 
> I now understand your first comment, sorry for misunderstanding.

Just FYI, Windows 7 Support was officially ended by Microsoft
https://www.microsoft.com/en-US/windows/windows-7-end-of-life-support-information

ppigazzini commented:
@Sopel97 give a try to msys2, it runs on windows7 :)

Sopel97 commented:
I've made it output into a zip archive, with source copied too. Improved logging and some naming.

I don't know how to check for desired bench easly, however. Any pointers?

Sopel97 commented:
Ok, I've added the bench verification. Bench is either passed by `--bench` parameter, or if not present it is extracted from `git log`. Is it ok to assume `--depth 1` log is enough for that in case of releases?

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,347 @@
+        raise Exception('The CRC check for {} failed.'.format(archive_path))
+
+    LOGGER.info('Created archive {}.'.format(archive_path))
+
```
In fishtest we use `subprocess.Popen()` to process in real time the output, IMO here is better to use always  `subprocess.run()`

```python
with open(os.devnull, 'wb') as dev_null:
    proc = subprocess.run(command, stderr = subprocess.PIPE, stdout = devnull, universal_newlines = True)
return int([line for line in proc.stderr.splitlines() if "Nodes searched" in line][0].split(": ")[1])
```

ppigazzini commented:
@Sopel97 I will prefer to process sequentially the tasks (i.e. `--concurrency 1`) but with `make -j <n_jobs>` to speedup the compiles.

Sopel97 commented:
I did it like this initially because bench is not parallelizable, so for large number of archs this is better. But I can see the number of jobs for make being higher than 1 useful too. I see two possible solutions:

1. add `--make-jobs` argument, effectively makes up to `concurrency * make_jobs` threads being used. This is the easiest but requires to user to worry about two parameters.
2. if `concurrency` is higher than `len(ARCHS)` then use the remaining threads for make jobs. This is IMO a better solution, but I'm not sure how to split the threads evenly. One solution could be to precompute the most equal distribution for each worker id and just use the same number of make jobs for each worker thread (if the thread id in the pool is accessible)

Sopel97 commented:
I've implemented option 2.

ppigazzini commented:
In my test sometimes some builds have failed with `--concurrency 15 --sde`, never happened with `--concurrency 1 --sde`.

Sopel97 commented:
OK. I may add a `--num-workers` option that overrides `NUM_WORKERS`

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,353 @@
+
+                archive.write(full_path, internal_path)
+
+    success = True
```
If I'm not wrong this could be shortened a bit

```python
    with ZipFile(archive_path, 'r') as archive:
        if archive.testzip() is not None:
            os.remove(archive_path)
            raise Exception('The CRC check for {} failed.'.format(archive_path))
```

Sopel97 replied:
can't remove the file while it's open. Could maybe do os.unlink but I think this is better.

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,330 @@
+    command = [stockfish, 'bench']
+    if SDE_PATH:
+        command = [SDE_PATH, '--'] + command
+
```
I reworked a bit also the list comprehension.
Anyway I used the comprehension because I'm experimenting with the cli, feel free to rework the code :)

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,353 @@
+                LOGGER.error('{}: Error running `{}`. Exit code: {}' \
+                            .format(arch, make_command_log, ret.returncode))
+            else:
+                src_file = os.path.join(src_dir, 'stockfish' + EXE_SUFFIX)
```
What about `strip`?
```python
subprocess.run( ['make', 'strip'], cwd=src_dir, stdout=stdout, stderr=STDERR)
```

Sopel97 replied:
right, I forgot about it

ppigazzini commented:
Snippet of fishtest code (written by @vondele), check if can be useful:

```python

def is_windows_64bit():
    if "PROCESSOR_ARCHITEW6432" in os.environ:
        return True
    return os.environ["PROCESSOR_ARCHITECTURE"].endswith("64")

def is_64bit():
    if IS_WINDOWS:
        return is_windows_64bit()
    return "64" in platform.architecture()[0]

def make_targets():
    """Parse the output of make help and extract the available targets"""
    p = subprocess.run(["make", "help"], stdout=subprocess.PIPE, universal_newlines=True)

    targets = []
    read_targets = False

    for line in p.stdout.splitlines():
        if "Supported compilers:" in line:
            read_targets = False
        if read_targets and len(line.split()) > 1:
            targets.append(line.split()[0])
        if "Supported archs:" in line:
            read_targets = True

    targets = [target for target in targets if "apple" not in target if "arm" not in target if "ppc" not in target]

    return targets


def gcc_props():
    """Parse the output of g++ -Q -march=native --help=target and extract the available properties"""
    p = subprocess.run(["g++", "-Q", "-march=native", "--help=target"], stdout=subprocess.PIPE, universal_newlines=True)

    flags = []
    arch = "None"
    for line in p.stdout.splitlines():
        if "[enabled]" in line:
            flags.append(line.split()[0])
        if "-march" in line and len(line.split()) == 2:
            arch = line.split()[1]

    return {"flags": flags, "arch": arch}


def find_arch_string():
    """Find the best ARCH=... string based on the cpu/g++ capabilities and Makefile targets"""

    targets = make_targets()
    props = gcc_props()

    if is_64bit():
        if (
            "-mavx512vnni" in props["flags"]
            and "-mavx512dq" in props["flags"]
            and "-mavx512f" in props["flags"]
            and "-mavx512bw" in props["flags"]
            and "-mavx512vl" in props["flags"]
            and "x86-64-vnni256" in targets
        ):
            res = "x86-64-vnni256"
        elif (
            "-mavx512f" in props["flags"]
            and "-mavx512bw" in props["flags"]
            and "x86-64-avx512" in targets
        ):
            res = "x86-64-avx512"
        elif (
            "-mbmi2" in props["flags"]
            and "x86-64-bmi2" in targets
            and props["arch"] not in ["znver1", "znver2"]
        ):
            res = "x86-64-bmi2"
        elif "-mavx2" in props["flags"] and "x86-64-avx2" in targets:
            res = "x86-64-avx2"
        elif (
            "-mpopcnt" in props["flags"]
            and "-msse4.1" in props["flags"]
            and "x86-64-modern" in targets
        ):
            res = "x86-64-modern"
        elif "-mssse3" in props["flags"] and "x86-64-ssse3" in targets:
            res = "x86-64-ssse3"
        elif (
            "-mpopcnt" in props["flags"]
            and "-msse3" in props["flags"]
            and "x86-64-sse3-popcnt" in targets
        ):
            res = "x86-64-sse3-popcnt"
        else:
            res = "x86-64"
    else:
        if (
            "-mpopcnt" in props["flags"]
            and "-msse4.1" in props["flags"]
            and "x86-32-sse41-popcnt" in targets
        ):
            res = "x86-32-sse41-popcnt"
        elif "-msse2" in props["flags"] and "x86-32-sse2" in targets:
            res = "x86-32-sse2"
        else:
            res = "x86-32"

    print("Available Makefile architecture targets: ", targets)
    print("Available g++/cpu properties: ", props)
    print("Determined the best architecture to be ", res)

    return "ARCH=" + res
```

Sopel97 commented:
We could use this to filter the targets in case SDE_PATH is not specified, but I don't think it's warranted for our use case given how much code it would add. We can revisit it later if needed.

ppigazzini commented:
`DEFAULT_ARCHS = make_targets()`

ppigazzini commented:
Error log
```
01:06:38 - ERROR - Exception: name 'make_command_log' is not defined
Traceback (most recent call last):
  File "make_releases.py", line 356, in run_task
    compile_target(
  File "make_releases.py", line 294, in compile_target
    .format(arch, make_command_log, ret.returncode))
NameError: name 'make_command_log' is not defined
```

ppigazzini commented:
@Sopel97 Let me know your targeted python release. `text` parameter in `subprocess.run()` was introduced with python 3.7.

Sopel97 commented:
I addressed the issues. Tagetted python is 3.6.3, replaced `text` with `encoding`.

Also replaced the default arch set to the one parsed from `make help`.

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,392 @@
+
+def get_bench_from_git_history(repo_dir):
+    command = ['git', '--git-dir', os.path.join(repo_dir, '.git'), 'log']
+    completed_process = subprocess.run(command, encoding='utf-8', capture_output=True, universal_newlines=True)
```
`capture_output` is python 3.7, if I'm not wrong only one of `universal_newlines=True` or `encoding='utf-8'` is required to have a string output.

`completed_process = subprocess.run(command, stdout=subprocess.PIPE, universal_newlines=True)`

Sopel97 replied:
right, I'll change this. But specifying `encoding` seems to still be required.

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,392 @@
+        def run_task(arch):
+            try:
+                with lock:
+                    thread_id = threading.current_thread().native_id
```
`threading.current_thread().native_id` python 3.8

It's running with `threading.current_thread().ident` but please double check, I made a monkey style bugfix :)

https://docs.python.org/3.6/library/threading.html#threading.Thread.ident

vondele commented:
I think there was a question on --depth 1 cloning, that wouldn't be enough to extract bench from git log. The 'No functional change' needs more. So, I would limit the depth but not to depth one. Maybe something like depth 25 or so.

Sopel97 commented:
hmm. might indeed be due to download failing. Definitely would be better to do it like you mention.

Sopel97 commented:
I added a commit which allows multiple tries per arch (`--max-tries`), default 1 try. This is a speculative change.

ppigazzini commented:
Problem solved with the single nn download :)
4 successful (and very fast) runs with `--concurrency 160 --max-num-workers 20 --sde "<sde_path>"`  

Your latest speculative commit broke all, though :(
```
14:31:15 - INFO - Created temporary directory cloned
14:31:15 - INFO - Cloning branch master from repository https://github.com/offic                                                                                                                               ial-stockfish/Stockfish.git into cloned
14:31:15 - INFO - Cloning command is: git clone --branch master --depth 25 https                                                                                                                               ://github.com/official-stockfish/Stockfish.git cloned
14:31:16 - INFO - Cloning finished.
14:31:16 - INFO - Executing `make net`.
Default net: nn-62ef826d1a6d.nnue
Downloading https://tests.stockfishchess.org/api/nn/nn-62ef826d1a6d.nnue
14:31:19 - INFO - Sucessfully downloaded the net.
14:31:19 - INFO - Expected bench is 4321677.
14:31:19 - INFO - Scheduling ARCH=x86-64-vnni512. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-vnni256. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-avx512. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-bmi2. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-avx2. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-sse41-popcnt. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-modern. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-ssse3. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64-sse3-popcnt. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-64. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-32-sse41-popcnt. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-32-sse2. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=x86-32. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=general-64. Try 1 out of 1
14:31:19 - INFO - Scheduling ARCH=general-32. Try 1 out of 1
14:31:19 - INFO - Failed building general-64 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-ssse3 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-bmi2 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-avx512 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-32-sse2 in an attempt 1 out of 1
14:31:19 - INFO - Failed building general-32 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-sse3-popcnt in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-vnni256 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-sse41-popcnt in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-vnni512 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-32 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-modern in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-64-avx2 in an attempt 1 out of 1
14:31:19 - INFO - Failed building x86-32-sse41-popcnt in an attempt 1 out of 1
14:31:19 - INFO - x86-64-vnni512: Failed
14:31:19 - INFO - x86-64-vnni256: Failed
14:31:19 - INFO - x86-64-avx512: Failed
14:31:19 - INFO - x86-64-bmi2: Failed
14:31:19 - INFO - x86-64-avx2: Failed
14:31:19 - INFO - x86-64-sse41-popcnt: Failed
14:31:19 - INFO - x86-64-modern: Failed
14:31:19 - INFO - x86-64-ssse3: Failed
14:31:19 - INFO - x86-64-sse3-popcnt: Failed
14:31:19 - INFO - x86-64: Failed
14:31:19 - INFO - x86-32-sse41-popcnt: Failed
14:31:19 - INFO - x86-32-sse2: Failed
14:31:19 - INFO - x86-32: Failed
14:31:19 - INFO - general-64: Failed
14:31:19 - INFO - general-32: Failed
14:31:19 - INFO - Removed temporary directory cloned
```

Sopel97 commented:
Yes, I commited a debug version by mistake, everything should be sorted now.

ppigazzini commented:
Works fine :)
(I haven't installed 32 bit gcc)
```
14:42:47 - INFO - Removed temporary directory cloned
14:42:47 - INFO - x86-64-vnni512: Success after 1 tries.
14:42:47 - INFO - x86-64-vnni256: Success after 1 tries.
14:42:47 - INFO - x86-64-avx512: Success after 1 tries.
14:42:47 - INFO - x86-64-bmi2: Success after 1 tries.
14:42:47 - INFO - x86-64-avx2: Success after 1 tries.
14:42:47 - INFO - x86-64-sse41-popcnt: Success after 1 tries.
14:42:47 - INFO - x86-64-modern: Success after 1 tries.
14:42:47 - INFO - x86-64-ssse3: Success after 1 tries.
14:42:47 - INFO - x86-64-sse3-popcnt: Success after 1 tries.
14:42:47 - INFO - x86-64: Success after 1 tries.
14:42:47 - INFO - x86-32-sse41-popcnt: Failed after 1 tries.
14:42:47 - INFO - x86-32-sse2: Failed after 1 tries.
14:42:47 - INFO - x86-32: Failed after 1 tries.
14:42:47 - INFO - general-64: Success after 1 tries.
14:42:47 - INFO - general-32: Failed after 1 tries.
```

snicolet commented:
> Now, I have one question. Should the .nnue file be included in the source? It is now.

I would say that including the nnue file in the source is probably not necessary, because

a) the nnue file is already embedded in the binaries
b) the sources folder is a copy of the Github sources, which do not include the nnue file.

(a) would make the size of the archive twice as big for no real benefit
(b) shows that users can compile SF easily as soon as they have a working Internet connection.

Sopel97 commented:
I agree with those points, excluded *.nnue from sources.

Lolligerhans commented on script/make_release.py:
```diff
@@ -0,0 +1,419 @@
+# For usage example use:
```
Add `#!/usr/bin/env python3` as first line?

snicolet commented:
I think we could merge this PR, if only to test the feature a bit before the rush for the next release.

ppigazzini commented:
@Sopel97 on Windows (either msys2-mingw64 shell or cmd setting path) the SDE bench string seems to be wrong: `\C:/Intel/sde-win/sde.exe" COMP=mingw -j8 -- ./stockfish bench > /dev/null`.

Here is a snippet of the verbose output on Windows:
```
python3 make_release.py --concurrency 8  --max-num-workers 2 --compiler mingw --verbose --sde C:/Intel/sde-win/sde.exe x86-64-avx512
...
Step 2/4. Running benchmark for pgo-build ...
\C:/Intel/sde-win/sde.exe" COMP=mingw -j8 -- ./stockfish bench > /dev/null
/bin/sh: -c: line 1: unexpected EOF while looking for matching `"'
/bin/sh: -c: line 2: syntax error: unexpected end of file
make: *** [Makefile:748: profile-build] Error 2
13:08:52 - ERROR - x86-64-avx512: Error running `make profile-build ARCH=x86-64-avx512 SDE_PATH="C:/Intel/sde-win/sde.exe" COMP=mingw -j8`. Exit code: 2
```

Hereis the same verbose output on WSL:
```
python3 make_release.py  --concurrency 8 --max-num-workers 2 --verbose --sde /home/usr00/__builder/sde-external-8.63.0-2021-01-18-lin/sde64 x86-64-avx512
...
Step 2/4. Running benchmark for pgo-build ...
"/home/usr00/__builder/sde-external-8.63.0-2021-01-18-lin/sde64" -- ./stockfish bench > /dev/null

Position: 1/47 (rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1)
```

By the way on Windows `make profile-build ARCH=x86-64-avx512 SDE_PATH="C:/Intel/sde-win/sde.exe" COMP=mingw -j8` works fine.

Sopel97 commented on script/make_release.py:
```diff
@@ -0,0 +1,463 @@
+            # it needs to be passed to the make script.
+            # There is no downside to always running under SDE if provided.
+            make_command = ['make', build_type, 'ARCH={}'.format(arch)]
+            if SDE_PATH:
+                make_command += ['SDE_PATH="{}"'.format(SDE_PATH)]
+            if compiler:
+                make_command += ['COMP={}'.format(compiler)]
+            if num_threads > 1:
+                make_command += ['-j{}'.format(num_threads)]
```
@ppigazzini probably the problem is somewhere here? The "escaping" on the `SDE_PATH` might be incorrect, and it ends up containing `COMP=...` stuff too.

ppigazzini replied:
Switching single quote - double quote at line 282 seems to solve the problem (???)
`make_command += ["SDE_PATH='{}'".format(SDE_PATH)]`

here is the ouput:
```
Step 2/4. Running benchmark for pgo-build ...
C:/Intel/sde-win/sde.exe -- ./stockfish.exe bench > /dev/null
```

Sopel97 commented:
So should be fixed now, but in the future we should look into proper escaping of the parameters.

ppigazzini commented on script/make_release.py:
```diff
@@ -0,0 +1,463 @@
+                LOGGER.info('{}: Finished build of {}.'.format(arch, src_file))
+                LOGGER.info('{}: Running `make strip`.'.format(arch))
+
+                subprocess.run(['make', 'strip'], cwd=src_dir, stdout=stdout, stderr=STDERR)
```
To work also on Windows (`EXE=stockfish.exe`):
```python
                make_command = ['make', 'strip']
                if compiler:
                    make_command += ['COMP={}'.format(compiler)]                
                subprocess.run(make_command, cwd=src_dir, stdout=stdout, stderr=STDERR)
```

vondele commented:
I would really like to see the building of these binaries to be part of our CI process. Only if we automate it will the script be really tested, not just the day of the release. Every commit could be built with this script and the binaries be made available automatically. If we just commit the script without automatically testing it, it is sub-optimal.

Sopel97 commented:
I agree. However I'm sure whether there's a nice way to save the artifacts from CI? and whether the size limits won't be a problem.

Sopel97 commented:
https://github.com/marketplace/actions/automatic-releases perhaps?

vondele commented:
something like that, indeed. There might be other options, I don't know how other projects are doing this, and I don't know github actions well, but this must be more or less a standard thing.

ppigazzini commented:
Some projects related to SF seems to have automatic build from CI:

https://github.com/ianfab/Fairy-Stockfish
https://github.com/niklasf/fishnet
https://github.com/cutechess/cutechess

snicolet commented:
Maybe we could ask @ianfab, @niklasf or @cutechess for a short crash course about automatic builds, then?

NightlyKing commented:
It works very well for me. However, I'm unable to compile 32 bit executables if using COMP=gcc (mingw64 and mingw32 shell). Using mingw as the comp poses no problem so with that it goes in 1 run. Although one has to remember to enable LTO for mingw. This way one loses no speed compared to COMP=gcc. COMP=mingw also supports static linking while also using LTO which doesn't work with COMP=gcc. I'm confident this is ready for the next release. AVX512 emulation is actually not too slow - it should work in CI. That is if we get to run it in CI (seems like SDE has problems in VMs).

This script would be great to have ready for SF 15 :)

xoto10 commented:
I suggest:
```
-                   help='The directory that shall contain the resulting binaries.')
+                   help='The directory that will contain the resulting binaries.')
```
I can't quote a rule, but shall doesn't seem right to me there. (btw, huge respect to you guys where English is not your native language!)

vondele commented:
There's even a wiki page https://en.wikipedia.org/wiki/Shall_and_will 

More seriously, I think the script is useful, but it needs to become part of a CI/CD setup, so that it is not just used for releases.

ppigazzini commented:
cmd to automate msys2 build:
1. minimal required msys2 packages
```cmd
C:\msys64\msys2_shell.cmd -defterm -msys2 -c "pacman -S --noconfirm --needed python git make mingw-w64-i686-gcc mingw-w64-x86_64-gcc mingw-w64-ucrt-x86_64-gcc mingw-w64-clang-x86_64-clang"
```
2.  Microsoft Visual C++ Runtime 32 bit builds
```cmd
C:\msys64\msys2_shell.cmd -defterm -mingw32 -here -c "python3 ./make_release.py --bench 4687476 --compiler mingw --keep-logs --concurrency 8 --max-num-workers 4 --prefix 14-win x86-32-sse41-popcnt x86-32-sse2 x86-32"
```
3. Microsoft Visual C++ Runtime 64 bit builds with SDE
```cmd
C:\msys64\msys2_shell.cmd -defterm -mingw64 -here -c "python3 ./make_release.py --bench 4687476 --compiler mingw --keep-logs --concurrency 8 --max-num-workers 4 --prefix 14-win --sde C:/Intel/sde-external-8.63.0-2021-01-18-win/sde.exe x86-64-vnni512 x86-64-vnni256 x86-64-avx512 x86-64-bmi2 x86-64-avx2 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3 x86-64-sse3-popcnt x86-64"
```
4. Universal C Runtime 64 bit builds with SDE (shipped with Windows 10 and later)
```cmd
C:\msys64\msys2_shell.cmd -defterm -ucrt64 -here -c "python3 ./make_release.py --bench 4687476 --compiler mingw --keep-logs --concurrency 8 --max-num-workers 4 --prefix 14-win --sde C:/Intel/sde-external-8.63.0-2021-01-18-win/sde.exe x86-64-vnni512 x86-64-vnni256 x86-64-avx512 x86-64-bmi2 x86-64-avx2 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3 x86-64-sse3-popcnt x86-64"
```
5. Clang 64 bit builds with SDE (requires dropping `-latomic` and adding `-static` in LDFLAGS, kernel: 'MINGW64_NT-10.0-19043')
```cmd
C:\msys64\msys2_shell.cmd -defterm -clang64 -here -c "python3 ./make_release.py --bench 4687476 --compiler mingw --keep-logs --concurrency 8 --max-num-workers 4 --prefix 14-win --sde C:/Intel/sde-external-8.63.0-2021-01-18-win/sde.exe x86-64-vnni512 x86-64-vnni256 x86-64-avx512 x86-64-bmi2 x86-64-avx2 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3 x86-64-sse3-popcnt x86-64"
```

see msys2 environments:
https://www.msys2.org/docs/environments/

PedanticHacker commented:
It would be nice if instead of using `format()`, the Python script should use f-strings.

PedanticHacker commented:
Currently, the cyclomatic complexity of this proposed script is somewhat high. `radon` reported this:
```
C:\Users\PedanticHacker\Desktop>radon cc "make_release.py" -a -s
make_release.py
    F 348:0 main - C (18)
    F 227:0 compile_target - B (10)
    F 326:0 get_possible_targets - B (10)
    F 160:0 pack_binary_with_source - B (8)
    F 219:0 run_bench - A (4)
    F 111:0 download_net - A (2)
    F 150:0 get_bench_from_git_history - A (2)
    C 120:0 DummyContextManager - A (2)
    C 131:0 TemporaryDirectory - A (2)
    M 136:4 TemporaryDirectory.__init__ - A (2)
    F 21:0 get_logger - A (1)
    F 87:0 remove_readonly - A (1)
    F 94:0 clone_stockfish - A (1)
    M 125:4 DummyContextManager.__enter__ - A (1)
    M 128:4 DummyContextManager.__exit__ - A (1)
    M 143:4 TemporaryDirectory.__enter__ - A (1)
    M 146:4 TemporaryDirectory.__exit__ - A (1)

17 blocks (classes, functions, methods) analyzed.
Average complexity: A (3.9411764705882355)
```
As you see, the most cyclomatically complex is the `main()` function, then `compile_target()`, `get_possible_targets()`, and `pack_binary_with_source()`. We should address these complexity issues.

PedanticHacker commented:
So, I shall make a separate PR to address the cyclomatic complexity issues and provide general improvements to this script.

vondele commented:
probably best to help work on the mentioned priorities, i.e. integrate this in a github actions based workflow that makes these artifacts available... that's the path towards getting this merged.

This is important stuff, but someone needs to bite the bullet.

PedanticHacker commented:
Guys, I found a guide that might help us: https://jcdan3.medium.com/how-to-run-python-scripts-in-your-github-actions-caccac969b34

PedanticHacker commented:
Anyway, GitHub Actions as a service **is** free, but there's a limit:
<!--StartFragment-->
GitHub plan | Total concurrent jobs | Maximum concurrent macOS jobs
-- | -- | --
Free | 20 | 5
<!--EndFragment-->
So, we should make the script be limited to spawn a maximum of 20 concurrent jobs only, as the above table says that this much is free.

PedanticHacker commented:
I've read all I could get on the topic of GitHub Actions. In short, I don't think we should take the GitHub Actions path at all. Isn't having Stockfish binaries on *abrok.eu/stockfish* good enough?

Anyway, in the long run, I do not see it's worth it, and that is because:
- GitHub gives only **2000 minutes per month** to run GitHub Actions.
- Those 2000 minutes reset every month, while **storage usage does not reset at all**.

The table below shows more information regarding this.
<!--StartFragment-->
Product | Storage | Minutes (per month)
-- | -- | --
GitHub Free | 500 MB | 2,000
GitHub Pro | 1 GB | 3,000
GitHub Free for organizations | 500 MB | 2,000
GitHub Team | 2 GB | 3,000
GitHub Enterprise Cloud | 50 GB | 50,000
<!--EndFragment-->

Now, let's look at a calculation of storage usage costs.

If one uses 3 GB of storage for 10 days in March and 12 GB for 21 days in the same month, the storage usage would be:
- 3 GB × 10 days × (24 hours per day) = 720 GB-hours
- 12 GB × 21 days × (24 hours per day) = 6,048 GB-hours
- 720 GB-hours + 6,048 GB-hours = 6,768 GB-hours
- 6,768 GB-hours ÷ (744 hours per month) = 9.0967  GB-months

So, at the end of the month, GitHub rounds the storage to the nearest MB. Therefore, the storage usage in March would be 9.097 GB.

Note that GitHub gives only 500 MB (0.5 GB) of storage for free, so unless someone will be paying for more GitHub Actions storage space every single month, we are out of luck.

Also, there are security risks involved in having a script. Namely, there exists a risk of script injections.

Let me explain this a little bit more. So, when creating workflows, custom actions, or composite actions, one should always consider whether the source code might execute as untrusted input from attackers. This **can** occur if an attacker adds malicious commands and specially crafted scripts to a context as strings of text. When a workflow runs, those strings might be interpreted as code which is then executed on the runner. Not good.

And since we are GitHub Actions noobs at this point... Yeah...

dav1312 commented:
> GitHub Actions usage is **free for** both **public repositories** and self-hosted runners. For **private repositories**, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account.

@PedanticHacker 
https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#:~:text=GitHub%20Actions%20usage%20is%20free%20for%20both%20public%20repositories

PedanticHacker commented:
Yes, GitHub Actions are **indeed** free, but with a twist. The twist being revealed in the below screenshot taken from my own GitHub account.

The screenshot has been taken while I was viewing this page: **https://github.com/settings/billing**. Please visit this page on your own while logged in to your GitHub account.
![image](https://user-images.githubusercontent.com/14254085/173209184-0639821e-52e3-4a67-b029-0441d4245415.png)
As you can see, I (and you probably as well) don't have unlimited number of minutes and unlimited storage. I (and you as well) have only 2000 minutes per month and 500 MB of storage available. That's the free portion you get. For more, pay montly.

So, my concern is whether 2000 minutes per month and 500 MB of storage is enough. @vondele?

dav1312 commented:
@PedanticHacker I don't think you are properly understanding what I'm trying to say.
Those limits are only for PRIVATE repositories.
Public repositories have free (unlimited) minutes and storage.

PedanticHacker commented:
Furthermore, the size of a Stockfish binary file is almost 50 MB nowadays, since NNUE is integrated by default. Now add 10 of such binary files, and you'll fill up the storage space rather quickly.

I was thinking, since GitHub Pro costs $4 per month or $48 per year, we could enable sponsoring the Stockfish project. With that money, we could pay for GitHub Pro. But then again, having 3,000 Actions minutes and 2 GB of storage as Pro doesn't seem such a vast upgrade to me. I don't know. I'm just putting my thoughts out there. I care about this.

PedanticHacker commented:
@dav1312, you say **free** and then put the word **unlimited** in the parentheses. Then why does my GitHub account have only 2,000 Actions minutes and 500 MB of "free" storage available?

dav1312 commented:
> why does my GitHub account have only 2,000 Actions minutes and 500 MB of "free" storage available?

@PedanticHacker  I already explained it twice, first time with bold letters and a second time with all caps, I'm not sure how can I be more clear.
I will repeat it one last time, **THOSE LIMITS are ONLY for PRIVATE REPOSITORIES**
Minutes and storage used in a **PUBLIC repository DON'T COUNT** towards those limits, they are FREE
More info: https://github.com/features/actions
![plans](https://user-images.githubusercontent.com/63931154/173227875-741b377a-c9bb-4fb2-8e2c-023162eb5f68.jpg)

PedanticHacker commented:
Okay, so how can we add this script to Github Actions? What is the procedure? As I quickly glanced at some instructions, I think we must define a specially crafted YAML file to achieve this.

PedanticHacker commented:
This discussion has deviated. I have contributed to that. I'm sorry.

Can we discuss what are we going to do with this Python script for making Stockfish releases? Improve upon it, reject it, accept it as it is? 🤔

bend-n commented:
Also, packages != releases. Packages are for docker images and stuff.

vondele commented:
I believe this can now be closed after merging https://github.com/official-stockfish/Stockfish/pull/4679 which provides similar functionality. Thanks!
